Alex/repository browser columns update

alexgomezlf opened the pull request:
- support columns in lf-repository-browser
- allow setting columns on lf-repository-browser
- for each column, lf-repository-browser will add a column
     - for each child entry, it will check the attributes property (Map) with key of "column.id", if it doesn't exist the value will be blank
     - allows sorting on each column
          - if you sort, the component will call for new data in the folder (like a refresh) but with the "columnOrderBy" parameter

Commit:
possible interface for columns support

Commit:
first commit

Commit:
update so that error and loading dialogs appear as before (using hidden instead of ngIf

Commit:
don't show header row when there are no "additional" columns

Commit:
use items as dataSource, not observable..

Commit:
verify ability to add/remove columns

Commit:
reset selected Values, reset selectable class based on isSelected property

rfulton-lf committed:
improved some css

Commit:
add sorting support

rfulton-lf committed:
fixed lint
removed lf-selection-list export from projects.ts

Commit:
use customservice

Commit:
Merge branch 'alex/repository-browser-columns-update' of https://github.com/Laserfiche/lf-ui-components into alex/repository-browser-columns-update

Commit:
fix dataSource update on scroll, sort, and new folder
sort behaves as new folder does
add refreshData event to tell lf-repository-browser that need to refresh folder

Commit:
update attributes property

Commit:
fix unit tests, set initial data on data source when data source is created after "listItems" is set

Commit:
add ability to check differenet columns

Commit:
resize columns (very rough)

rfulton-lf committed:
did work today :)

Commit:
move resize bar to column separator
make it so first resize doesn't shift by 10+ pixels due to original scaling issue

rfulton-lf committed:
outer-aligned resize handles

rfulton-lf committed:
added ellipses for data cells, still working on headers

rfulton-lf committed:
truncated long header cells, no ellipses (yet)

rfulton-lf committed:
made resizing continuous, added minimum width (100px)

Commit:
add ability to keep same table size when adding column that will scale to less than the minimum size

Commit:
remove extra forEach

Commit:
add TODO comment

shiyuanzhao-lf committed:
update README.md

Ke-Jin-Laserfiche committed:
set select width

Ke-Jin-Laserfiche committed:
behave like web client (always resize left column)

Ke-Jin-Laserfiche committed:
call setTableResize when listItems is set

shiyuanzhao-lf committed:
Merge branch 'alex/resize-columns' of https://github.com/Laserfiche/lf-ui-components into alex/resize-columns

Ke-Jin-Laserfiche committed:
refactor, fix multiple repo-browser same page

Ke-Jin-Laserfiche committed:
use percentage

Ke-Jin-Laserfiche committed:
fix jumpiness

Ke-Jin-Laserfiche committed:
don't sort on resize

Ke-Jin-Laserfiche committed:
add cursor on both side of the column divider

Commit:
calculate select width

Commit:
don't resize based on table size, remove hack for resize

Commit:
allow width percentage re-calculation if the width of container changes

Ke-Jin-Laserfiche committed:
Merge pull request #102 from Laserfiche/ke/resize-column

resize columns

rfulton-lf committed:
stored the customized column widths in local storage

Ke-Jin-Laserfiche committed:
set table width

Ke-Jin-Laserfiche committed:
fix unit test

Commit:
fix localstorage interface, remove unnecessary code

Commit:
Merge branch 'alex/resize-columns' of https://github.com/Laserfiche/lf-ui-components into alex/resize-columns

Commit:
rough version of using css grid

Commit:
rough version of using css grid

Commit:
get rid of currentWidth property

Commit:
make percentage sum to container, overflow-x scroll not working

Commit:
simplify offset calculation

Commit:
user tr as grid instead of table, sticky row not working

Commit:
fix sticky column headers

Commit:
update selection and calculation of indices

Commit:
only update column widths on mouseup

Commit:
use wrapper div to make resizer more readable

Ke-Jin-Laserfiche committed:
use scrollLeft to calculate the distanced scrolled

Ke-Jin-Laserfiche committed:
after initialization, save column widths as pixels

Commit:
use clientx, not pagex

Ke-Jin-Laserfiche committed:
resizePosition account for min column width

Ke-Jin-Laserfiche committed:
don't add duplicate name column, fix resize position, clean up

Commit:
add ability to pass in custom name column reference

Commit:
don't need to use resIzeObserver to get containerWidth

Commit:
simplify, move datasource to separate class

Commit:
fix lint

Ke-Jin-Laserfiche committed:
add test

Ke-Jin-Laserfiche committed:
Merge branch 'alex/use-grid-css' of https://github.com/Laserfiche/lf-ui-components into alex/use-grid-css

Ke-Jin-Laserfiche committed:
fix lint

Ke-Jin-Laserfiche committed:
fix tests

Ke-Jin-Laserfiche committed:
remove containerwidth

Commit:
add arrow staying in place

Commit:
change isLoading logic back to initial implementation

Commit:
fix selction list and resizing on sorting

Ke-Jin-Laserfiche committed:
grid-resize cursor is col-resize

Ke-Jin-Laserfiche committed:
initialize width no smaller than minWidth

Ke-Jin-Laserfiche committed:
Merge branch 'alex/use-grid-css' of https://github.com/Laserfiche/lf-ui-components into alex/use-grid-css

Ke-Jin-Laserfiche committed:
fix css

Ke-Jin-Laserfiche committed:
update dpcumentation

Commit:
don't update data as often

Commit:
update initial column sizing, fill width if only one column, else set default

Commit:
update styles, change additionalColumns from setter to function

Commit:
make dataSource more readable
make item height configurable
fix tab in/out of component

Ke-Jin-Laserfiche committed:
fix lint

Ke-Jin-Laserfiche committed:
fix tests

Commit:
refactor selection list, move resize to directive

Commit:
Merge branch 'alex/repository-browser-columns-update' of https://github.com/Laserfiche/lf-ui-components into alex/repository-browser-columns-update

Ke-Jin-Laserfiche committed:
fix tests

Commit:
update selction list to use minWidth

Commit:
update repository browser documentation

Commit:
fix lint, add internal

Commit:
fix column initialization before initAsync, fix border css issue in firefox

Ke-Jin-Laserfiche commented on projects/lf-documentation/src/app/lf-repository-browser-documentation/demo-repo-service.ts:
```diff
@@ -307,8 +327,45 @@ export class DemoRepoService implements LfTreeNodeService {
+          return this.compare(aVal.toLowerCase(), (bVal as string).toLowerCase(), isAsc);
+        } else {
+          // err -- not valid?
+          // or just do straight comparison?
```
do a console.warn and return 0?

Ke-Jin-Laserfiche commented on projects/lf-documentation/src/app/lf-repository-browser-documentation/lf-repository-browser-documentation.component.html:
```diff
@@ -146,66 +202,47 @@
     <td>
-      Opens the selected nodes. If one container is selected it will open that folder in the UI, if multiple items or an item that is not a container
-      is selected, then this function will just emit the entryOpened event with the selected nodes.
+      Opens the selected nodes. If one container is selected it will open that folder in the UI, if multiple items or an
```
if one container is selected, will it also emit the entryOpened event?

Ke-Jin-Laserfiche commented on projects/ui-components/lf-repository-browser/lf-repository-browser.component.spec.ts:
```diff
@@ -11,7 +11,8 @@ import { MatDialog } from '@angular/material/dialog';
+import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
+import { ColumnDef, LfSelectionListComponent, LfSelectionListModule, RepositoryBrowserData } from '@laserfiche/lf-ui-components/lf-selection-list';
 
 const rootTreeNode: LfTreeNode = {
```
add in test that if a column is not resizeable, should not be able to resize
and if a column is not sortable, should not be able to sort
if a column is sortable, should sort

Ke-Jin-Laserfiche replied:
test just call the sort function
add if no column header, use defaultWidth

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/resize-column.directive.ts:
```diff
@@ -0,0 +1,87 @@
+      const table = this.renderer.parentNode(thead);
+      this.viewport = this.renderer.parentNode(table);
+      this.divresizer = this.renderer.createElement('div');
+      this.divresizer!.classList.add('repository-browser-resize-overlay');
```
if not resizable, should we add a css so that the cursor for resize-handle is 'not-allowed' or something that indicates that the column is not resizable?

Commit:
do not use user settings when not showing header in lf-repository-browser, add documentation

Commit:
only set defaultWidth if no header

Commit:
fix filter and refresh when item is selected,
still have bug if you select something while filtered, it clears any not currently displayed, but dhould be fixed in demo service

Commit:
only emit selected values that exist

Commit:
update styles

Commit:
update styles, remove filter in documentation

Commit:
fix build in production

Commit:
make number column not resizable

Ke-Jin-Laserfiche commented on projects/lf-documentation/src/app/lf-repository-browser-documentation/demo-repo-service.ts:
```diff
@@ -307,8 +319,42 @@ export class DemoRepoService implements LfTreeNodeService {
+        } else {
+          // No value or unsupported value
+        }
+        return 0;
```
move return 0 inside else

Ke-Jin-Laserfiche commented on projects/lf-documentation/src/app/lf-repository-browser-documentation/lf-repository-browser-documentation.component.html:
```diff
@@ -23,6 +24,11 @@
     <p>repoBrowser.addEventListener('entrySelected', onElementNodeSelected(event));</p>
     <p>await repoBrowser.initAsync(treeProvidersObject);</p>
+    <p>const create: ColumnDef = 	&#123; id: 'creationTime', displayName: 'Creation Date', defaultWidth: '40%', minWidth: 100, resizable: true, sortable: true &#125;;</p>
+    <p>const templateName : ColumnDef = &#123; id: 'templateName', displayName: 'Template Name', defaultWidth: '40%', minWidth: 200, resizable: true &#125;;</p>
```
just show one

Ke-Jin-Laserfiche commented on projects/lf-documentation/src/app/lf-repository-browser-documentation/lf-repository-browser-documentation.component.html:
```diff
@@ -146,66 +201,47 @@
+    <td>focuses the first item in the lf-repository-browser</td>
   </tr>
   <tr>
-    <td>isSelectable</td>
```
read only? all be properties?

Ke-Jin-Laserfiche commented on projects/styles/lf-laserfiche-lite/sass/_typography-lf.scss:
```diff
@@ -50,14 +50,22 @@ $sec-info-color: #6D6D6D;
+    color: $error-color;
+  }
 }
 
```
add to ms styles

Ke-Jin-Laserfiche commented on projects/ui-components/lf-repository-browser/lf-repository-browser.component.html:
```diff
@@ -1,34 +1,32 @@
+      <div *ngIf="getIcons(item).length > 0" class="fixed-size-icon-container  lf-no-user-select">
+        <img *ngFor="let icon of getIcons(item)" src="{{ icon }}" alt="" class="lf-entry-icon" />
+      </div>
+      <div title="{{ item?.name }}" class="lf-ellipsis-label">{{ item?.name }}</div>
```
add lf-no-user-select

Ke-Jin-Laserfiche commented on projects/ui-components/lf-repository-browser/lf-repository-browser.component.spec.ts:
```diff
@@ -63,17 +65,19 @@ describe('LfRepositoryBrowserComponent', () => {
-    changeRefMock = jasmine.createSpyObj('changeDetectorRef', ['detectChanges']);
-    localizeServiceMock = jasmine.createSpyObj('localizeion', {
+    changeRefMock = jasmine.createSpyObj('ref', ['detectChanges']);
+    localizeServiceMock = jasmine.createSpyObj('localizationService', {
```
remove

Ke-Jin-Laserfiche commented on projects/ui-components/lf-repository-browser/lf-repository-browser.component.spec.ts:
```diff
@@ -51,7 +53,7 @@ const moduleDef: TestModuleMetadata = {
     LfSelectionListModule,
   ],
-  declarations: [LfRepositoryBrowserComponent, LfBreadcrumbsComponent, LfLoaderComponent],
+  declarations: [LfRepositoryBrowserComponent, LfBreadcrumbsComponent, LfLoaderComponent, LfSelectionListComponent],
```
remove LfSelectionListComponent

Ke-Jin-Laserfiche commented on projects/ui-components/lf-repository-browser/lf-repository-browser.component.spec.ts:
```diff
@@ -505,6 +527,173 @@ describe('LfRepositoryBrowserComponent', () => {
   });
+
+  it('setSelectedNodesAsync should set selectedNode', async () => {
+    // TODO: this will fail because component.entryList is undefined
```
refactor. remove comment

Ke-Jin-Laserfiche commented on projects/ui-components/lf-repository-browser/lf-repository-browser.component.spec.ts:
```diff
@@ -505,6 +527,173 @@ describe('LfRepositoryBrowserComponent', () => {
+    // need to reset create columneDef to have the defaultWidth because the test
+    const propIdCreateDate: string = 'create_date';
+    const createDateInitialWidth = '35%';
+    const create : ColumnDef = { id: propIdCreateDate, displayName: 'Creation Date', defaultWidth: createDateInitialWidth, minWidth: 100, resizable: true, sortable: true  };
```
try remove it

Ke-Jin-Laserfiche commented on projects/ui-components/lf-repository-browser/lf-repository-browser.component.spec.ts:
```diff
@@ -505,6 +527,173 @@ describe('LfRepositoryBrowserComponent', () => {
+    expect(createDateWidth).toBe(createDateActualWidth);
+  }));
+
+  it('if name is provided in columnToDisplay, use the defaultWidth there', fakeAsync(async () => {
```
rename the title: if name is passed into setAdditionalColumnsToDisplay...

Ke-Jin-Laserfiche commented on projects/ui-components/lf-repository-browser/lf-repository-browser.component.spec.ts:
```diff
@@ -505,6 +527,173 @@ describe('LfRepositoryBrowserComponent', () => {
+    const createDateInitialWidth = '35%';
+    const create : ColumnDef = { id: propIdCreateDate, displayName: 'Creation Date', defaultWidth: createDateInitialWidth, minWidth: 100, resizable: true, sortable: true  };
+    const name : ColumnDef = { id: 'name', displayName: 'Name', defaultWidth: '65%', minWidth: 100, resizable: true, sortable: true };
+    const initialNameColumnWidth = parseFloat('65%')*containerWidth/100;
```
65% to 80%

Ke-Jin-Laserfiche commented on projects/ui-components/lf-repository-browser/lf-repository-browser.component.spec.ts:
```diff
@@ -505,6 +527,173 @@ describe('LfRepositoryBrowserComponent', () => {
+    const resizeHandleEls = Array.from(
+      document.getElementsByClassName('resize-handle')
+    );
+    const initialClientX = (resizeHandleEls[0].getBoundingClientRect().right + resizeHandleEls[0].getBoundingClientRect().left)/2 ;
```
refactor this to the directive.spec.ts
don't need to test localstorage
check column width is updated
or store function is called
mock the directive here

Ke-Jin-Laserfiche commented on projects/ui-components/lf-repository-browser/lf-repository-browser.component.ts:
```diff
@@ -520,11 +688,13 @@ export class LfRepositoryBrowserComponent implements OnDestroy {
     this.lastCalledPage = this.nextPage;
-    const dataPage = await this.treeNodeService.getFolderChildrenAsync(parentEntry, this.nextPage);
+    const sortState = this.columnOrderBy;
+    const dataPage = await this.treeNodeService.getFolderChildrenAsync(parentEntry, this.nextPage, sortState);
```
type as LfTreeNodePage

Ke-Jin-Laserfiche commented on projects/ui-components/lf-repository-browser/lf-repository-browser.component.ts:
```diff
@@ -1,36 +1,128 @@
-
+export class LfRepositoryBrowserComponent implements OnDestroy, AfterViewInit {
+  /**
+   * read-only property to get current breadcrumbs
```
test comment pops up in npm package and types package

Ke-Jin-Laserfiche commented on projects/ui-components/lf-repository-browser/lf-repository-browser.component.ts:
```diff
@@ -1,36 +1,128 @@
+    if (this.entryList) {
+      const nameColumnDef = cols.find((col) => col.id === 'name');
+      let fillLastColumn: boolean;
+      if (cols.length > 1 || (cols.length === 1 && cols[0].id !== 'name')) {
```
simply this logic

Ke-Jin-Laserfiche commented on projects/ui-components/lf-repository-browser/lf-repository-browser.component.ts:
```diff
@@ -477,16 +629,32 @@ export class LfRepositoryBrowserComponent implements OnDestroy {
+          this.ref.detectChanges();
+          const selected = this.selectedItems;
+          this.entryList?.clearSelectedValues();
+          const selectableValues = await this.mapTreeNodesToLfSelectableAsync(selected ?? []);
```
clear it up

Ke-Jin-Laserfiche commented on projects/ui-components/lf-repository-browser/lf-repository-browser.component.ts:
```diff
@@ -520,11 +688,13 @@ export class LfRepositoryBrowserComponent implements OnDestroy {
     this.nextPage = dataPage.nextPage;
     const page = dataPage.page;
     selectablePage = await this.mapTreeNodesToLfSelectableAsync(page);
+    this.isLoading = false;
```
check if still needed

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list-data-source.ts:
```diff
@@ -0,0 +1,81 @@
+import { Subscription, BehaviorSubject, Observable, Subject } from 'rxjs';
+import { ILfSelectable } from '@laserfiche/lf-ui-components/shared';
+
+const PAGESIZE = 50;
```
a way to change the pageSize?

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list-data-source.ts:
```diff
@@ -0,0 +1,81 @@
+    this.viewport.setTotalContentSize(this.itemSize * this._data.length);
+    if (!existingData || existingData.length === 0) {
+      this.resetView();
+    }
```
reconsider

Commit:
refactor reset selectedItems in refresh

Commit:
simplify how columns are set

Commit:
remove isLoading in updateFolderChildrenAsync, tested no longer needed

Commit:
add types to untyped variables

Commit:
add no user select to columns

Commit:
add error-text color to microsoft style sheet
remove mulriple columns documentation

Ke-Jin-Laserfiche commented on projects/ui-components/lf-checklist/items/items.component.css:
```diff
@@ -24,10 +24,6 @@
 
-mat-checkbox.lf-checkbox {
-  width: 100%;
-}
```
should this be fixed by removing viewEncapsulation in the other PR #106 ?

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list-data-source.ts:
```diff
@@ -0,0 +1,81 @@
+    });
+  }
+
+  private readonly visibleData: BehaviorSubject<ILfSelectable[]> = new BehaviorSubject<ILfSelectable[]>([]);
```
move this up

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list-data-source.ts:
```diff
@@ -0,0 +1,81 @@
+    super();
+    this._data = initialData;
+    this.viewport.setTotalContentSize(this.itemSize * initialData.length);
+    this.visibleData.next(this._data.slice(0, PAGESIZE));
```
define this.dataStart (0) and this.dataEnd (pageSize) earlier

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list-types.ts:
```diff
@@ -0,0 +1,21 @@
+/** @internal */
+export interface SelectedItemEvent {
+  selected: ILfSelectable;
+  selectedItems: ILfSelectable[] | undefined;
```
check if this is visible in the types package

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.css:
```diff
@@ -1,12 +1,203 @@
+  outline: var(--lf-menu-item-active-outline-color, #1783d6) auto 2px !important;
+  background-color: var(--lf-menu-item-hover-color, #f5fbff);
+}
+.item-holder.fill-multiselect {
```
remove

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.css:
```diff
@@ -1,12 +1,203 @@
+  padding: 0 0 0 15px;
+}
+
+::ng-deep .lf-ellipsis-label .mat-sort-header-content {
```
review css, be more specific

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.css:
```diff
@@ -1,12 +1,203 @@
+}
+
+:host .lf-table-selection-list th.mat-header-cell {
+  overflow: visible !important;
```
see if we can review it

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.css:
```diff
@@ -1,12 +1,203 @@
+  background: linear-gradient(#000, #000) no-repeat center/2px 100%;
+}
+
+::ng-deep .lf-table-selection-list th.placeholder {
```
double check

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.css:
```diff
@@ -1,12 +1,203 @@
+  cursor: col-resize;
+  opacity: 0;
+  width: 15px;
+  right: -17.5px;
```
add a comment on where to get the numbers from

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.html:
```diff
@@ -1,27 +1,142 @@
+          <td mat-cell *matCellDef="let item">
+            <div class="lf-name-col-container">
+              <ng-container *ngIf="listItemRef; else defaultItem"
+                ><ng-container *ngIf="item">
```
check if the second ng-container is needed

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.html:
```diff
@@ -1,27 +1,142 @@
+            <!-- <mat-checkbox (change)="$event ? masterToggle() : null"
+                                  [checked]="selection.hasValue() && isAllSelected()"
+                                  [indeterminate]="selection.hasValue() && !isAllSelected()">
+                    </mat-checkbox> -->
```
remove for now

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.html:
```diff
@@ -1,27 +1,142 @@
+              color="primary"
+              (click)="onCheckboxClicked($event)"
+              disableRipple
+              [checked]="row.isSelected"
```
remove duplicated [checked]

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.html:
```diff
@@ -1,27 +1,142 @@
+        </ng-container>
+      </ng-container>
+    </ng-container>
+    <ng-container matColumnDef="item-description">
```
remove item-description

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.ts:
```diff
@@ -2,78 +2,188 @@ import { FocusMonitor, FocusOrigin } from '@angular/cdk/a11y';
+    this.items = items;
+    if (this.dataSource) {
+      if (this.dataSource.allData.length < 1 && items.length > 0) {
+        if (this._showHeader === true) {
```
double check this._showHeader is always set

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.ts:
```diff
@@ -2,78 +2,188 @@ import { FocusMonitor, FocusOrigin } from '@angular/cdk/a11y';
+  }
+  set columnOrderBy(orderBy: ColumnOrderBy | undefined) {
+    if (orderBy?.columnId && this.allColumnDefs.find((c) => c.id === orderBy.columnId)) {
+      this.sort!.sort({ id: orderBy?.columnId, start: orderBy?.isDesc ? 'desc' : 'asc', disableClear: true });
```
this.sort should exist. if doesn't, console.warn

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.ts:
```diff
@@ -2,78 +2,188 @@ import { FocusMonitor, FocusOrigin } from '@angular/cdk/a11y';
+      this.setInitialWidth();
+    } else {
+      this.setDefaultWidths();
+    }
```
refractor

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.ts:
```diff
@@ -84,28 +194,48 @@ export class LfSelectionListComponent implements AfterViewInit, OnDestroy {
+  }
+
+  placeholderWhen(index: number, _: any) {
+    return index == 0;
```
===

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.ts:
```diff
@@ -84,28 +194,48 @@ export class LfSelectionListComponent implements AfterViewInit, OnDestroy {
 
+  resetSelectedValues(list: ILfSelectable[]) {
+    this.selectable.resetSelectedValues(list);
+  }
```
remove

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.ts:
```diff
@@ -115,19 +245,20 @@ export class LfSelectionListComponent implements AfterViewInit, OnDestroy {
+      this.selectable.onItemClicked(event, option, this.items);
     }
     this.currentFocusIndex = index;
+    // TODO not focusing
```
remove

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.ts:
```diff
@@ -163,31 +294,19 @@ export class LfSelectionListComponent implements AfterViewInit, OnDestroy {
   onFocused(index: number) {
-    this.itemFocused.emit(this.listItems[index].value);
+    this.currentFocusIndex = index;
+    // TODO this is
```
remove  todo

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.ts:
```diff
@@ -227,13 +352,81 @@ export class LfSelectionListComponent implements AfterViewInit, OnDestroy {
+      if (this.allColumnDefs.length > 0) {
+        const repositoryBrowserData: RepositoryBrowserData | undefined = this.getRepositoryBrowserData();
+        const tableEl = Array.from(
+          this.viewport!.elementRef.nativeElement.getElementsByClassName('lf-table-selection-list')
```
is this this.matTable?

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.ts:
```diff
@@ -227,13 +352,81 @@ export class LfSelectionListComponent implements AfterViewInit, OnDestroy {
+        this.ref.detectChanges();
+
+        // TODO is there a better way to do this
+        if (!(this.allColumnDefs[this.allColumnDefs.length - 1]?.defaultWidth === 'auto')) {
```
simplify this logic

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/resize-column.directive.ts:
```diff
@@ -0,0 +1,87 @@
+      const thead = this.renderer.parentNode(row);
+      const table = this.renderer.parentNode(thead);
+      this.viewport = this.renderer.parentNode(table);
+      this.divresizer = this.renderer.createElement('div');
```
rename better

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/resize-column.directive.ts:
```diff
@@ -0,0 +1,87 @@
+    event.preventDefault();
+    event.stopPropagation();
+    if (!this.viewport) {
+      return;
```
warn

Ke-Jin-Laserfiche commented on projects/ui-components/shared/LfSelectable.ts:
```diff
@@ -32,6 +32,21 @@ export class Selectable {
     this.clearAllSelectedItems(list);
   }
 
+  resetSelectedValues(list: ILfSelectable[]) {
```
remove

Commit:
remove resetSelectedValues, clarify naming in resizeColumnDirective

Commit:
clean code, refactor into readable variables

Ke-Jin-Laserfiche committed:
add tests

Commit:
update css, remove unnecessary html

Commit:
Merge branch 'alex/repository-browser-columns-update' of https://github.com/Laserfiche/lf-ui-components into alex/repository-browser-columns-update

Commit:
update css, remove important when not needed

Commit:
update inputs to be snake case, add input for page_size

Commit:
make datasource internal

Commit:
fix lint

Commit:
Merge branch '14.x' of https://github.com/Laserfiche/lf-ui-components into alex/repository-browser-columns-update

Commit:
update documentation

Ke-Jin-Laserfiche committed:
refactor tests

Ke-Jin-Laserfiche committed:
fix import

Ke-Jin-Laserfiche committed:
add tests for setAdditionalColumnToDisplay

Ke-Jin-Laserfiche committed:
add test for always show headers

Ke-Jin-Laserfiche committed:
remove localization mock

Ke-Jin-Laserfiche committed:
refractor

Ke-Jin-Laserfiche committed:
remove extra component reference

Ke-Jin-Laserfiche committed:
fix lint

Ke-Jin-Laserfiche commented on projects/lf-documentation/src/app/lf-repository-browser-documentation/lf-repository-browser-documentation.component.html:
```diff
@@ -23,6 +24,9 @@
     <p>await repoBrowser.initAsync(treeProvidersObject);</p>
+    <p>const create: ColumnDef = 	&#123; id: 'creationTime', displayName: 'Creation Date', defaultWidth: '40%', minWidth: 100, resizable: true, sortable: true &#125;;</p>
+    <p>const columns: ColumnDef[] = [create];</p>
+    <p>repoBrowser.setAdditionalColumns(columns);</p>
```
setAdditionalColumnsToDisplay

Ke-Jin-Laserfiche commented on projects/ui-components/lf-repository-browser/lf-repository-browser.component.spec.ts:
```diff
@@ -85,10 +119,12 @@ describe('LfRepositoryBrowserComponent', () => {
   let matDialogMock: MatDialog;
   let ngZoneMock: NgZone;
 
+  const containerWidth = 500;
```
remove

Ke-Jin-Laserfiche commented on projects/ui-components/lf-repository-browser/lf-repository-browser.component.spec.ts:
```diff
@@ -469,42 +505,59 @@ describe('LfRepositoryBrowserComponent', () => {
+
+  it('if name column is passed into setAdditionalColumnsToDisplay, set the name column to width to be be the default width', () => {
+
+    const nameCol: ColumnDef = { id: 'name', displayName: 'Name', defaultWidth: '80%', minWidth: 100, resizable: true, sortable: true };
```
format document

Ke-Jin-Laserfiche commented on projects/ui-components/lf-repository-browser/lf-repository-browser.component.ts:
```diff
@@ -1,36 +1,146 @@
+   * Property to set whether or not to always show header. By default will only show header if there is more than one column
+   */
+  @Input()
+  set always_show_header(showHeader: boolean | undefined | string) {
```
test in sample app setting the input will trigger change detection

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/lf-selection-list.component.spec.ts:
```diff
@@ -1,14 +1,23 @@
+import { ILfSelectable, ItemWithId, PropertyValue } from '@laserfiche/lf-ui-components/shared';
+import { ColumnDef, SelectedItemEvent } from './lf-selection-list-types';
+import { LfSelectionListComponent, RepositoryBrowserData } from './lf-selection-list.component';
+const propIdCreateDate: string = 'create_date';
```
add new line

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/resize-column.directive.spec.ts:
```diff
@@ -0,0 +1,184 @@
+  widthChanged: number = 100;
+  gridTemplateColumns: string = `${this.widthChanged}px 100px 100px`;
+  onWidthChange(width: number) {
+    console.log('width changed', width);
```
remove console.log

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/resize-column.directive.spec.ts:
```diff
@@ -0,0 +1,184 @@
+
+describe('ResizeColumnDirective', () => {
+  let fixture: ComponentFixture<TestComponent>;
+  beforeEach(async () => {
```
test if fResizeColumn is false

Ke-Jin-Laserfiche replied:
hard to test ngOnInit is only called once in the directive. This should not be a problem because the directive is only for internal use. Will skip this test for now

Ke-Jin-Laserfiche commented on projects/ui-components/lf-selection-list/resize-column.directive.spec.ts:
```diff
@@ -0,0 +1,184 @@
+      });
+      const mouseMoveEvent = new MouseEvent('mousemove', {
+        clientX: initialClientRight + moveX,
+        movementX: moveX,
```
add test for minWidth

Ke-Jin-Laserfiche committed:
add change log, add more test

Ke-Jin-Laserfiche committed:
remove console.log

Ke-Jin-Laserfiche committed:
remove extra function

Ke-Jin-Laserfiche committed:
fix documentation

Commit:
update repository browser documentation and interface

Commit:
don't double submit the entrySelected if still no entries selected
